### PR TITLE
8274975: Add micro benchmark: ALIBABA selective store use case

### PR DESCRIPTION
**Pull request by JoshuaZhuwj**
I separate my implementation of "compress" API into several patches for easy review.
This micro benchmark is extracted from the real use case inside in-memory databases: selectiveStore.
Test scenario: UseAVX=3; thread number = 8; conflict data percentage: 20% (that means 20% of mask bits are true)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274975](https://bugs.openjdk.java.net/browse/JDK-8274975): Add micro benchmark: ALIBABA selective store use case

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - Committer) ⚠️ Review applies to c4d51d6f921b90c3e59949bb67464e0b5949de25

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-vector pull/149/head:pull/149` \
`$ git checkout pull/149`

Update a local copy of the PR: \
`$ git checkout pull/149` \
`$ git pull https://git.openjdk.java.net/panama-vector pull/149/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 149`

View PR using the GUI difftool: \
`$ git pr show -t 149`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-vector/pull/149.diff">https://git.openjdk.java.net/panama-vector/pull/149.diff</a>

</details>

### CODE REVIEW

File: test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+      VectorMask<Integer> mask = av.compare(VectorOperators.NE, bv);
+      IntVector dv = cv.compress(mask);
+      dv.intoArray(tld.conflict_vector, tld.conflict_vector_cnt, mask.prefixMask());
+      tld.conflict_vector_cnt += mask.trueCount();
+    }
+  }
+
+  @Benchmark
+  @Threads(8)
+  public void selectiveStore_verify(ThreadLocalData tld) {
```
Comment (theRealELiu) [inline test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java:163]: Is it necessary to benchmark the correctness?

Comment (JoshuaZhuwj) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java:163]: Thanks Eric for your review :-)
This function was imported to ensure compress API's correctness during my development process.
I will remove this function as your suggestion after all dependent patches are merged.

**Comment by JoshuaZhuwj** [comment]
Using vector-to-vector compress, the benchmark performance result:
image

File: test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+  @Benchmark
+  @Threads(8)
+  public void selectiveStore_vector_128(ThreadLocalData tld) {
+    tld.conflict_vector_cnt = 0;
+    for (int i = 0; i < ARRAY_LENGTH; i += INT_128_SPECIES.length()) {
+      IntVector av = IntVector.fromArray(INT_128_SPECIES, tld.input1, i);
+      IntVector bv = IntVector.fromArray(INT_128_SPECIES, tld.input2, i);
+      IntVector cv = IntVector.fromArray(INT_128_SPECIES, tld.index, i);
+      VectorMask<Integer> mask = av.compare(VectorOperators.NE, bv);
```
Comment (jatin-bhateja) [inline test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java:109]: Can we mover vector compare out of benchmarking loop to reduce the noise.

Comment (JoshuaZhuwj) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java:109]: Thanks Jatin for your review.
This test I added is placed together with other benchmarks extracted from user cases.
The mask computation inside the loop is the essential logic in real sceanario.
We can add micro benchmarks in test/micro/org/openjdk/bench/jdk/incubator/vector/operation and then evaluate performance for compression / expanding in single API level.

Comment (jatin-bhateja) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java:109]: Ok got it, thanks! 
BTW can you also post the AVX2 numbers just to see how does it compare against scalar version when direct compress/expand instructions are not supported by target.

Comment (JoshuaZhuwj) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/bigdata/SelectiveStore.java:109]: FYI
image

**Comment by theRealELiu** [comment]
> How about the performance for byte and short types? Do you have performance data comparing Java implementation against with scalar version? As some platforms may not support the compress/expand operation well. E.g., SVE doesn’t support compress on byte and short types. AVX512 maybe the same without vbmi2 extension.

I think the result of AVX2 could reflect the same issue. It's a shame that SVE misses the direct compress instruction for byte and short types.

**Comment by JoshuaZhuwj** [comment]
> How about the performance for byte and short types? Do you have performance data comparing Java implementation against with scalar version? As some platforms may not support the compress/expand operation well. E.g., SVE doesn’t support compress on byte and short types. AVX512 maybe the same without vbmi2 extension.

Eric, on platforms without hardware support, like AVX2 or Neon, vector-to-vector compression could be a composition of mask -> shuffle -> rearrange. A precomputed permutation table could be leveraged to lookup shuffle from mask. John Rose elaborated the algorithm how to compute a compression shuffle from a mask. See design discussion at https://github.com/openjdk/panama-vector/pull/115

**Comment by JoshuaZhuwj** [comment]
Thanks for all your review comments. I update the benchmark accordingly.

**Pull request closed without merging**